### PR TITLE
Adds delete batch job 

### PR DESCRIPTION
**Pull request by jpengst**
# Summary
Create batch job to delete parent objects by uploading a csv of oids.

# Related Ticket
[#1407](https://app.zenhub.com/workspaces/yul-dc-5e50083561915b11fc9e5850/issues/yalelibrary/yul-dc/1407)

# Video

https://user-images.githubusercontent.com/36549923/127907834-242fdfc3-3716-4aac-aba1-7490488eb320.mp4

### CODE REVIEW

File: app/controllers/child_objects_controller.rb [hunk 1]
```diff
@@ -70,7 +70,14 @@ def destroy
...
     def set_child_object
-      @child_object = ChildObject.find(params[:id]) if params[:id]
+      begin
+        @child_object = ChildObject.find(params[:id]) if params[:id]
+      rescue ActiveRecord::RecordNotFound
+        respond_to do |format|
+          format.html { redirect_to parent_objects_url, notice: "Child object, oid: #{params[:id]}, was not found in local instance." }
+          format.json { head :no_content }
+        end
+      end
```
Comment (orangewolf) [inline app/controllers/child_objects_controller.rb]: No real reason to raise an exception here. 

```suggestion
      @child_object = ChildObject.find_by(id: params[:id]) if params[:id]
      if @child_object.blank?
        respond_to do |format|
          format.html { redirect_to parent_objects_url, notice: "Child object, oid: #{params[:id]}, was not found in local instance." }
          format.json { head :no_content }
        end
      end
```

File: app/controllers/parent_objects_controller.rb [hunk 2]
```diff
@@ -141,7 +141,14 @@ def invalidate_admin_set_edit
 
     # Use callbacks to share common setup or constraints between actions.
     def set_parent_object
-      @parent_object = ParentObject.find(params[:id])
+      begin
+        @parent_object = ParentObject.find(params[:id])
```
Comment (orangewolf) [inline app/controllers/parent_objects_controller.rb]: see above

File: app/jobs/remove_from_metadata_cloud_csv_job.rb [hunk 3]
```diff
@@ -0,0 +1,13 @@
+# frozen_string_literal: true
+
+class RemoveFromMetadataCloudCsvJob < ApplicationJob
+  queue_as :default
+
+  def default_priority
+    -50
```
Comment (orangewolf) [inline app/jobs/remove_from_metadata_cloud_csv_job.rb]: this is a very hi propriety, is that what you mean to do here?

File: app/models/batch_process.rb [hunk 4]
```diff
@@ -106,6 +106,33 @@ def create_parent_objects_from_oids(oids, metadata_sources, adminset_keys)
       end
     end
   end
+
+  def delete_parent_object(oids, metadata_sources, adminset_keys)
```
Comment (orangewolf) [inline app/models/batch_process.rb]: use admin_set_key, just to be consistent please

File: app/models/batch_process.rb [hunk 5]
```diff
@@ -106,6 +106,33 @@ def create_parent_objects_from_oids(oids, metadata_sources, adminset_keys)
...
+    admin_set_hash = {}
+    oids.zip(metadata_sources, adminset_keys).each_with_index do |record, index|
+      oid, metadata_source, adminset_key = record
+      admin_set = admin_set_hash[adminset_key]
```
Comment (orangewolf) [inline app/models/batch_process.rb]: all this admin_set questing belongs in its own method

File: app/models/batch_process.rb [hunk 6]
```diff
@@ -106,6 +106,33 @@ def create_parent_objects_from_oids(oids, metadata_sources, adminset_keys)
...
+        admin_set = AdminSet.find_by_key(adminset_key)
+        admin_set_hash[adminset_key] = admin_set
+      end
+      if admin_set.nil?
+        batch_processing_event("Skipping row [#{index + 2}] with unknown admin set [#{adminset_key}] for parent: #{oid}", 'Skipped Row')
+        next
+      end
+      next unless user_edit_permission(index, admin_set, oid)
+
+      if ParentObject.where(oid: oid).count.negative?
```
Comment (orangewolf) [inline app/models/batch_process.rb]: what does `ParentPbject.destroy` do? is it different from `po = ParentObject.find_by(oid: oid); po.destroy if po.present?`

File: app/models/batch_process.rb [hunk 7]
```diff
@@ -121,6 +148,12 @@ def refresh_metadata_cloud_csv
     create_parent_objects_from_oids(oids, metadata_sources, admin_sets)
   end
 
+  def remove_from_metadata_cloud_csv
+    metadata_sources = parsed_csv.entries.map { |r| r['source'] }
```
Comment (orangewolf) [inline app/models/batch_process.rb]: its expensive to run through these columns twice. can we do a single pass instead?

**Review by orangewolf — CHANGES_REQUESTED** [review]

File: app/jobs/remove_from_metadata_cloud_csv_job.rb [hunk 8]
```diff
@@ -0,0 +1,13 @@
+# frozen_string_literal: true
+
+class RemoveFromMetadataCloudCsvJob < ApplicationJob
```
Comment (mikeapp) [inline app/jobs/remove_from_metadata_cloud_csv_job.rb]: Can we drop the MetadataCloud reference here?  It's a misnomer in the other job as well.  Something like "DeleteObjectsJob" would be fine.

Comment (jpengst) [reply app/jobs/remove_from_metadata_cloud_csv_job.rb]: Yep, no problem! Thanks for the feedback

File: app/controllers/child_objects_controller.rb (hunk 1 above)
Comment (K8Sewell) [reply app/controllers/child_objects_controller.rb]: We attempted to use the `find_by` method but that resulted in the error below.  So we stuck with what we had.  Open to trying again though.

image

File: app/controllers/parent_objects_controller.rb (hunk 2 above)
Comment (K8Sewell) [reply app/controllers/parent_objects_controller.rb]: same as above

File: app/models/batch_process.rb [hunk 9]
```diff
@@ -84,45 +84,68 @@ def created_file_name
...
-    create_parent_objects_from_oids(oids, metadata_sources, admin_sets)
+    parsed_csv.each_with_index do |row, index|
+      oid = row['oid']
+      metadata_source = row['source']
+      admin_set = editable_admin_set(row['admin_set'], oid, index)
+      next unless admin_set
+
+      parent_object = ParentObject.find_or_initialize_by(oid: oid)
+      # Only runs on newly created parent objects
+      next unless parent_object.new_record?
```
[the remark this answers is not in the document]
Comment (K8Sewell) [reply app/models/batch_process.rb]: Just updated this and should be through the pipeline soon.  Everything passed locally.

**Review by martinlovell — APPROVED** [review]

**Review by orangewolf — APPROVED** [review]

**Pull request merged**